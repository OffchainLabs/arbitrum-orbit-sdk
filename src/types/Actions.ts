import { Address } from 'viem';
import { Prettify } from './utils';

/**
 * Actions require contract address, but as part of decorators, the address might have been passed already to the decorator.
 *
 * If the address was passed to the decorator, it's now optional (we still allow overrides of the address per action).
 * If the action doesn't have any other parameters beside the contract address, then parameters can either be { contract: address } or void
 */
export type ActionParameters<Args, ContractName extends string, Curried extends boolean> = Prettify<
  Curried extends false
    ? Args & { [key in ContractName]: Address }
    : Args extends Record<string, never>
    ?
        | {
            [key in ContractName]: Address;
          }
        | void
    : Args & {
        [key in ContractName]?: Address;
      }
>;
