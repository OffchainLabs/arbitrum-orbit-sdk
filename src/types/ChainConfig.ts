import { Address } from 'viem';

export type ChainConfigArbitrumParams = {
  EnableArbOS: boolean;
  AllowDebugPrecompiles: boolean;
  InitialArbOSVersion: number;
  InitialChainOwner: Address;
  DataAvailabilityCommittee: boolean;
  GenesisBlockNum: number;
  MaxCodeSize: number;
  MaxInitCodeSize: number;
};

export type ChainConfig = {
  chainId: number;
  homesteadBlock: number;
  daoForkBlock: null;
  daoForkSupport: boolean;
  eip150Block: number;
  eip150Hash: string;
  eip155Block: number;
  eip158Block: number;
  byzantiumBlock: number;
  constantinopleBlock: number;
  petersburgBlock: number;
  istanbulBlock: number;
  muirGlacierBlock: number;
  berlinBlock: number;
  londonBlock: number;
  clique: {
    period: number;
    epoch: number;
  };
  arbitrum: ChainConfigArbitrumParams;
};
