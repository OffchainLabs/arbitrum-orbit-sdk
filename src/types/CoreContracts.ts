import { Address } from 'viem';

export type CoreContracts = {
  rollup: Address;
  nativeToken: Address;
  inbox: Address;
  outbox: Address;
  rollupEventInbox: Address;
  challengeManager: Address;
  adminProxy: Address;
  sequencerInbox: Address;
  bridge: Address;
  upgradeExecutor: Address;
  validatorUtils?: Address;
  validatorWalletCreator: Address;
  deployedAtBlockNumber: number;
};
