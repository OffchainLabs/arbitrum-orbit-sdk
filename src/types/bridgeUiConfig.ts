import { Chain, PublicClient, Transport } from 'viem';

/**
 * Bridge UI configuration information for adding a new test arbitrum chain
 */
export type BridgeUiConfig = {
  /**
   * Basic chain information
   */
  chainInfo: {
    chainName: string;
    chainId: number;
    parentChainId: number;
    rpcUrl: string;
    explorerUrl: string;
    nativeToken?: string;
  };
  /**
   * Core protocol contract addresses
   */
  coreContracts: {
    rollup: string;
    inbox: string;
    outbox: string;
    sequencerInbox: string;
    bridge: string;
    nativeToken?: string;
  };
  /**
   * Token bridge contract addresses for both parent chain (L2) and Orbit chain (L3)
   */
  tokenBridgeContracts: {
    /** Parent chain (L2) contract addresses */
    l2Contracts: {
      customGateway: string;
      multicall: string;
      proxyAdmin: string;
      router: string;
      standardGateway: string;
      weth: string;
      wethGateway: string;
    };
    /** Orbit chain (L3) contract addresses */
    l3Contracts: {
      customGateway: string;
      multicall: string;
      proxyAdmin: string;
      router: string;
      standardGateway: string;
      weth: string;
      wethGateway: string;
    };
  };
};

/**
 * Parameters for retrieving bridge UI configuration
 */
export type GetBridgeUiConfigParams<TChain extends Chain | undefined> = {
  parentChain: TChain;
  deploymentTxHash: `0x${string}`;
  chainName?: string;
  rpcUrl?: string;
  explorerUrl?: string;
  parentChainPublicClient?: PublicClient<Transport, TChain>;
};
