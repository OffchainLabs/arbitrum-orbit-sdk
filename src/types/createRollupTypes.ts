import { Address, GetFunctionArgs } from 'viem';

import { rollupCreator } from '../contracts';

export type CreateRollupFunctionInputs = GetFunctionArgs<
  typeof rollupCreator.abi,
  'createRollup'
>['args'];

type RequiredKeys = 'config' | 'batchPoster' | 'validators';

export type CreateRollupParams = Pick<CreateRollupFunctionInputs[0], RequiredKeys> &
  Partial<Omit<CreateRollupFunctionInputs[0], RequiredKeys>>;

export type WithRollupCreatorAddressOverride<T> = T & {
  /**
   * Specifies a custom address for the RollupCreator. By default, the address will be automatically detected based on the provided chain.
   */
  rollupCreatorAddressOverride?: Address;
};
