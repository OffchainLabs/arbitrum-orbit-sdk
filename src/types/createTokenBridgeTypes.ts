import { Address } from 'viem';

export type WithTokenBridgeCreatorAddressOverride<T> = T & {
  /**
   * Specifies a custom address for the TokenBridgeCreator. By default, the address will be automatically detected based on the provided chain.
   */
  tokenBridgeCreatorAddressOverride?: Address;
};
