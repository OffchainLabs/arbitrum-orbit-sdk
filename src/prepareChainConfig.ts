import { ChainConfig, ChainConfigArbitrumParams } from './types/ChainConfig';

export const defaults = {
  homesteadBlock: 0,
  daoForkBlock: null,
  daoForkSupport: true,
  eip150Block: 0,
  eip150Hash: '0x0000000000000000000000000000000000000000000000000000000000000000',
  eip155Block: 0,
  eip158Block: 0,
  byzantiumBlock: 0,
  constantinopleBlock: 0,
  petersburgBlock: 0,
  istanbulBlock: 0,
  muirGlacierBlock: 0,
  berlinBlock: 0,
  londonBlock: 0,
  clique: {
    period: 0,
    epoch: 0,
  },
  arbitrum: {
    EnableArbOS: true,
    AllowDebugPrecompiles: false,
    DataAvailabilityCommittee: false,
    InitialArbOSVersion: 20,
    GenesisBlockNum: 0,
    MaxCodeSize: 24_576,
    MaxInitCodeSize: 49_152,
  },
};

export type PrepareChainConfigParams = Pick<ChainConfig, 'chainId'> &
  Partial<Omit<ChainConfig, 'chainId' | 'arbitrum'>> & {
    arbitrum: Pick<ChainConfigArbitrumParams, 'InitialChainOwner'> &
      Partial<Omit<ChainConfigArbitrumParams, 'InitialChainOwner'>>;
  };

export function prepareChainConfig(params: PrepareChainConfigParams): ChainConfig {
  return {
    ...defaults,
    ...params,
    clique: { ...defaults.clique, ...params.clique },
    arbitrum: { ...defaults.arbitrum, ...params.arbitrum },
  };
}
