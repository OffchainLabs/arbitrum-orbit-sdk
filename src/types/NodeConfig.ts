import { ChainConfig } from './ChainConfig';

export type NodeConfigChainInfoJson = [
  {
    'chain-id': number;
    'parent-chain-id': number;
    'parent-chain-is-arbitrum': boolean;
    'chain-name': string;
    'chain-config': ChainConfig;
    'rollup': {
      'bridge': string;
      'inbox': string;
      'sequencer-inbox': string;
      'rollup': string;
      'validator-utils': string;
      'validator-wallet-creator': string;
      'deployed-at': number;
    };
  },
];

export type NodeConfigDataAvailabilityRpcAggregatorBackendsJson = [
  {
    url: string;
    pubkey: string;
    signermask: number;
  },
];
