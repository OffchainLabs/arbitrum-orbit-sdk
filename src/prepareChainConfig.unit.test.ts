import { it, expect } from 'vitest';

import { ChainConfig } from './types/ChainConfig';
import { prepareChainConfig } from './prepareChainConfig';

const chainId = 69_420;
const vitalik: `0x${string}` = '0xd8da6bf26964af9d7eed9e03e53415d37aa96045';

it('creates chain config with defaults', () => {
  const params = {
    chainId,
    arbitrum: {
      InitialChainOwner: vitalik,
    },
  };

  expect(prepareChainConfig(params)).toMatchSnapshot();
});

it('creates chain config with custom params', () => {
  const params: ChainConfig = {
    chainId,
    homesteadBlock: 1,
    daoForkBlock: null,
    daoForkSupport: false,
    eip150Block: 1,
    eip150Hash: '0x1100000000000000000000000000000000000000000000000000000000000000',
    eip155Block: 1,
    eip158Block: 1,
    byzantiumBlock: 1,
    constantinopleBlock: 1,
    petersburgBlock: 1,
    istanbulBlock: 1,
    muirGlacierBlock: 1,
    berlinBlock: 1,
    londonBlock: 1,
    clique: {
      period: 1,
      epoch: 1,
    },
    arbitrum: {
      EnableArbOS: false,
      AllowDebugPrecompiles: true,
      DataAvailabilityCommittee: true,
      InitialArbOSVersion: 20,
      InitialChainOwner: vitalik,
      GenesisBlockNum: 1,
      MaxCodeSize: 40 * 1024,
      MaxInitCodeSize: 80 * 1024,
    },
  };

  expect(prepareChainConfig(params)).toMatchSnapshot();
});
