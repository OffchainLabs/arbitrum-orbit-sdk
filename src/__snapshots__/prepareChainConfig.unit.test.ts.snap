// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`creates chain config with custom params 1`] = `
{
  "arbitrum": {
    "AllowDebugPrecompiles": true,
    "DataAvailabilityCommittee": true,
    "EnableArbOS": false,
    "GenesisBlockNum": 1,
    "InitialArbOSVersion": 20,
    "InitialChainOwner": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
    "MaxCodeSize": 40960,
    "MaxInitCodeSize": 81920,
  },
  "berlinBlock": 1,
  "byzantiumBlock": 1,
  "chainId": 69420,
  "clique": {
    "epoch": 1,
    "period": 1,
  },
  "constantinopleBlock": 1,
  "daoForkBlock": null,
  "daoForkSupport": false,
  "eip150Block": 1,
  "eip150Hash": "0x1100000000000000000000000000000000000000000000000000000000000000",
  "eip155Block": 1,
  "eip158Block": 1,
  "homesteadBlock": 1,
  "istanbulBlock": 1,
  "londonBlock": 1,
  "muirGlacierBlock": 1,
  "petersburgBlock": 1,
}
`;

exports[`creates chain config with defaults 1`] = `
{
  "arbitrum": {
    "AllowDebugPrecompiles": false,
    "DataAvailabilityCommittee": false,
    "EnableArbOS": true,
    "GenesisBlockNum": 0,
    "InitialArbOSVersion": 20,
    "InitialChainOwner": "0xd8da6bf26964af9d7eed9e03e53415d37aa96045",
    "MaxCodeSize": 24576,
    "MaxInitCodeSize": 49152,
  },
  "berlinBlock": 0,
  "byzantiumBlock": 0,
  "chainId": 69420,
  "clique": {
    "epoch": 0,
    "period": 0,
  },
  "constantinopleBlock": 0,
  "daoForkBlock": null,
  "daoForkSupport": true,
  "eip150Block": 0,
  "eip150Hash": "0x0000000000000000000000000000000000000000000000000000000000000000",
  "eip155Block": 0,
  "eip158Block": 0,
  "homesteadBlock": 0,
  "istanbulBlock": 0,
  "londonBlock": 0,
  "muirGlacierBlock": 0,
  "petersburgBlock": 0,
}
`;
